FEATURE: Refresh the rawContentMode

mficzel opened the pull request:
The raw content mode is implemented freshly and now only relies on the nodetype-configuration for rendering and not requiring a fusion integration. By default the raw mode will render all inline editable properties, previews of Images and all childnodes. This allows content first workflows with Neos or headless applications.

![Home_-_Neos_Demo_Site](https://user-images.githubusercontent.com/1309380/175993575-986fed63-9c9d-4a06-ac95-d51296b96530.png)

It is possible to provide a special raw-mode rendering for nodetype by creating a prototype with the name of the nodetype + '.RawContentMode' as shown in the example below.

```
prototype(Neos.Demo:Content.Columns.Two.RawContentMode) < prototype(Neos.Neos:ContentComponent) {
    renderer = afx`
        <div style="display:grid; grid-template-columns: 1fr 1fr; grid-gap: 16px;">
            <div><Neos.Neos:RawContent.NodeChildren @context.node={q(node).children('column0').get(0)} /></div>
            <div><Neos.Neos:RawContent.NodeChildren @context.node={q(node).children('column1').get(0)} /></div>
        </div>
    `
}
```

Resolves: #3705

mficzel committed:
FEATURE: Refresh the rawContentMode

Raw content mode is implemented fresh and only relies on the nodetypes for rendering without using the node rendering in live mode at all.
By default the raw mode will render all inline editable properties together previews of Images.

It is possible to provide a special raw-mode rendering for nodetype by creating a prototype with the name of the nodetype + '.Raw'

mficzel committed:
TASK: Delete old raw content templates

mhsdesign commented on Neos.Neos/Resources/Private/Fusion/RawContent/NodeProperties.fusion:
```diff
@@ -0,0 +1,16 @@
+    node = null
+
+    renderer = afx`
+        <Neos.Fusion:Loop items={props.node.nodeType.fullConfiguration.properties} itemKey="propertyName" itemName="propertyConfiguration">
```
this getter https://github.com/neos/neos-development-collection/blob/fd83b396c9d5641bf1c983a1cdd449fa38214552/Neos.ContentRepository/Classes/Domain/Model/NodeType.php#L336 

> Should only be used internally.
> Instead, use the hasConfiguration()/getConfiguration() methods to check/retrieve single configuration values.

but `getConfiguration("properties")` wont work in eel.
i know we are the core and allowed to use internal stuff, but if someone looks at it he wont know that and mistake it as api...

what do you think?

mficzel replied:
Nice catch adjusted ... turns out `nodeType.properties` already does all i need. 

See https://github.com/neos/neos-development-collection/blob/fd83b396c9d5641bf1c983a1cdd449fa38214552/Neos.ContentRepository/Classes/Domain/Model/NodeType.php#L424

mficzel committed:
TASK: Use `nodeType.properties` instead of  `nodeType.fullConfiguration.properties`

mficzel committed:
TASK: Restructure rendering

mficzel committed:
TASK: Adjust styles of raw content mode

mhsdesign commented on Neos.Neos/Resources/Private/Fusion/RawContent/Node.fusion:
```diff
@@ -0,0 +1,45 @@
+            renderer = Neos.Fusion:Renderer {
+                @context.node = ${props.node}
+                type = ${props.node.nodeType.name + '.RawContentMode'}
+                element.node = ${props.node}
```
i dont see the value of saving `node = ${node}` in the component ... and injecting node magically. `@context` is for most people scary enough already this could blow them away xD

mhsdesign commented:
btw while on this, logically the current 'fusion include dependency graph' is a bit odd: `DefaultFusion.fusion`  renders the [`rawContent` fusion path](https://github.com/neos/neos-development-collection/blob/17248b57e8717695b54a63ea09b07b97a0bcd371/Neos.Neos/Resources/Private/Fusion/DefaultFusion.fusion#L24)
so `RawContentMode.fusion` should be included from `DefaultFusion.fusion` and not from the `Root.fusion` right?

while thinking even more about it, why dont we just `import **/*.fusion` in `Root.fusion` and dont try to handle the dependencies our selves?

mficzel committed:
TASK: Use dynamic list for rendering and ensure that nodes that show no properties at least render the label

mficzel committed:
TASK: Simplify rendering and always show translated node labels

mficzel committed:
TASK: Tweak css

Sebobo commented on Neos.Neos/Resources/Private/Fusion/RawContent/NodeProperties.fusion:
```diff
@@ -0,0 +1,18 @@
+        <dl class="neos-raw-content-properties" @if={!String.isBlank(this.content)}>
+            <Neos.Fusion:Loop items={props.node.nodeType.properties} itemKey="propertyName" itemName="propertyConfiguration">
+                <Neos.Fusion:Fragment @if.isInlineEditable={propertyConfiguration.ui.inlineEditable == true} class="property">
+                    <dt>{propertyName}:</dt>
```
Use readable label if available

```suggestion
                    <dt>{propertyConfiguration.ui.label || propertyName}:</dt>
```

Sebobo commented on Neos.Neos/Resources/Private/Fusion/RawContent/NodeProperties.fusion:
```diff
@@ -0,0 +1,18 @@
+                    <dd><Neos.Neos:Editable property={propertyName} /></dd>
+                </Neos.Fusion:Fragment>
+                <Neos.Fusion:Fragment @if.isImage={propertyConfiguration.type == 'Neos\Media\Domain\Model\ImageInterface' && q(props.node).property(propertyName)}>
+                    <dt>{propertyName}:</dt>
```
Use readable label if available

```suggestion
                    <dt>{propertyConfiguration.ui.label || propertyName}:</dt>
```

mficzel committed:
Update Neos.Neos/Resources/Private/Fusion/RawContent/NodeProperties.fusion

Co-authored-by: Sebastian Helzle <sebastian@helzle.net>

mficzel committed:
Update Neos.Neos/Resources/Private/Fusion/RawContent/NodeProperties.fusion

Co-authored-by: Sebastian Helzle <sebastian@helzle.net>

Sebobo commented on Neos.Neos/Resources/Private/Fusion/RawContent/NodeProperties.fusion:
```diff
@@ -0,0 +1,18 @@
+                </Neos.Fusion:Fragment>
+                <Neos.Fusion:Fragment @if.isImage={propertyConfiguration.type == 'Neos\Media\Domain\Model\ImageInterface' && q(props.node).property(propertyName)}>
+                    <dt>{propertyConfiguration.ui.label || propertyName}:</dt>
+                    <dd><Neos.Neos:ImageTag asset={q(props.node).property(propertyName)} maximumWidth="200" /></dd>
```
Make sure images don't get to high

```suggestion
                    <dd><Neos.Neos:ImageTag asset={q(props.node).property(propertyName)} maximumHeight={200} maximumWidth={200} /></dd>
```

mficzel committed:
Update Neos.Neos/Resources/Private/Fusion/RawContent/NodeProperties.fusion

Co-authored-by: Sebastian Helzle <sebastian@helzle.net>

mficzel commented:
add NodeType Icons (if feasible) and hidden state

mficzel committed:
TASK: Improve rendering

- Show icon
- Render separator behind name and property
- Translate property names

mficzel committed:
TASK: Add visibility informations to each node

rolandschuetz commented:
Here are three ideas how the design could look like. My personal favorit is "light".

## Light
<img width="2031" alt="Raw Content Mode - light" src="https://user-images.githubusercontent.com/735982/168825396-6e5fce8b-80f3-457c-a5de-73206086bcae.png">

## Tagged
<img width="2031" alt="Raw Content Mode - tagged" src="https://user-images.githubusercontent.com/735982/168825406-a8007666-3cfd-4043-896d-0ad04c796a2e.png">

## Technical
<img width="2031" alt="Raw Content Mode - technical" src="https://user-images.githubusercontent.com/735982/168825422-c7adf223-b857-45de-8da8-2f2cd8d86a7c.png">

Sebobo commented:
@rolandschuetz I like "light" also the best, thx for the mockups. 

Don't we waste a bit of space with the spacing around the document node?

For me this view could be a bit more part of the ui itself and wouldn't need to "distance" itself so much in space and style.

rolandschuetz commented:
@Sebobo As a general rule in design text shouldn't be longer than 60ch to be nicely readable. We are already way above that. The same goes for images, is the would take up the whote horizontal width they would just use to mich of the browser height.

I think restricting it actually makes is better to use

mficzel commented:
"Light" is also my favorite. A really nice one. Can you adjust the css directly ob the pr?

mficzel committed:
TASK: Add raw content mode adjustments for the contents from Neos NodeTypes.ColumnLayouts

rolandschuetz commented:
@Sebobo @mficzel Martin wanted the UI to look different from the Neos UI. Otherwise, it would be quite dark and very tech looking. But sure, we could also include the Neos backend styling and just use this. I'm fine with both solutions.

Our current Neos UI does not have a concept of generic nesting (z-index). In the inspector the darker outside layer is behind the tab content, but these are just two dimensions. In the content area we have 5+ of those nestings, so we can't really work with color anymore. The UI doesn't have a concept to visualize this. In most interfaces this is done with a) shadows or b) lighter colors in the front.

I would be curious, what style most ppl prefer for raw content editing. I'm happy to implement which ever is preferred.

grebaldi commented:
@Sebobo 
> I just don't want to go so far away in terms of styling & design language from the Neos UI itself. The color can be different but it should use the same principles, fonts, spacing, etc. IMO.

@rolandschuetz 
> [...] Otherwise, it would be quite dark and very tech looking.

I agree with @Sebobo's statement that it's perfectly fine to go for a lighter color theme (We wanted to provide a light theme at some point anyway - so, the new RawContentMode could lay the groundwork for that). We should still keep things consistent in terms of font, spacing, etc. These are just a few design primitives (all described here: https://github.com/neos/neos-ui/blob/master/packages/build-essentials/src/styles/styleConstants.js).

Personally, I'm in favor of the color theme as outlined in the screenshots above :)

iseries commented:
I also like the light theme but the outer gap could be a bit smaller.

mficzel committed:
TASK: Improve styles by combining input from roland schuetz and default spacings and sizes from the neos ui

mficzel commented:
@rolandschuetz i tried to combine your ideas from the "light" suggestion with default sizes and paddings from other places in the ui.

Sebobo commented:
@mficzel can you post another screenshot for comparison?

mficzel commented:
@Sebobo ... i updated the screenshot in the pr description

Sebobo commented:
Love it

mficzel commented on Neos.Neos/Resources/Private/Fusion/RawContent/Node.fusion:
```diff
@@ -0,0 +1,40 @@
+                    <Neos.Fusion:Renderer
+                        @context.node={props.node}
+                        type={props.node.nodeType.name + '.RawContentMode'}
+                        element.node={props.node}
```
@mhsdesign would you prefer to go via @context.node in all places. Would make sense to me aswell.

mhsdesign replied:
Okay maybe i dindt explain it correctly ^^

the above renders a prototype like:

`Vendor.Site:Headline.RawContentMode`

and the integration can look like this due to the magic `element.node={props.node}`:

```
prototype(Vendor.Site:Headline.RawContentMode) < prototype(Neos.Neos:ContentComponent) {
    renderer = afx`{q(props.node).property("title")}`
}
```

but id expect to have to write the line `node={node}`:

```
prototype(Vendor.Site:Headline.RawContentMode) < prototype(Neos.Neos:ContentComponent) {
    node = ${node}
    renderer = afx`{q(props.node).property("title")}`
}
```


of course shitty examples.

IMO its not needed as we dont do it with the ContentCase either ...

mhsdesign replied:
TLTR: Just remove that line ... no?

mficzel committed:
TASK: Communicate node via context

Communicating the current node via context or prop can lead to confusion. I adjusted the code to only use @context.node now
to have a single source of truth what the current node is.

In addition the default coding style was applied to the fusion files.

mficzel commented:
@mhsdesign i removed those lines and adjusted the code to get the current node always from @context.node. This makes sense as this is the way contentElementWrapping and contentCase work. Also it was confusing to keep @context.node and props.node apart.

mhsdesign reviewed (approved):
@mficzel sorry for the extra work - i think we talked a bit past each other but everything seems fine now! (by studying / reading) but will try this out sometime!

> In addition the default coding style was applied to the fusion files.

My eyes hurt after reading the diff 😂 i think that was the revenge xD

rolandschuetz committed:
TASK: UX improvements

rolandschuetz commented:
@mficzel I like the implementation. I would add a little padding on the document node to show that this is already an element. And I would restrict the size an image can get to less than 100%.

<img width="1912" alt="Bildschirmfoto 2022-07-18 um 01 42 40" src="https://user-images.githubusercontent.com/735982/179429559-767e0046-34bd-4934-b980-f78840d9a5c1.png">


In my local setup, the two column layout it build with _Neos.Demo:Content.Columns.Two_  and so I do not have two columns visually.

I would also propose to show the property names with _String.firstLetterToUpperCase(string)_, what do you think?

mficzel committed:
TASK: Improve rendering

1. Add global padding to always show the border of the document node
2. Apply String.firstLetterToUpperCase to property names

mficzel committed:
TASK: Adjust rawContentMode preview for column contents

mficzel commented:
@rolandschuetz Thanks for the review. Added global padding and String.firstLetterToUpperCase for property names. Do not really get what you mean regarding the image width. Behaves nicely for me.

To verify the multicolumn preview you can checkout the neos-demo pr for that: https://github.com/neos/Neos.Demo/pull/128

mficzel committed:
Merge branch 'feature/refreshRawContentMode' into feature/refreshRawContentMode

mficzel committed:
TASK: Use preset "'Neos.Media.Browser:Thumbnail'" for rendering the images in raw content mode to use already existing resources

mficzel committed:
TASK: Move padding to outer container so select border of document node is visible

rolandschuetz commented:
Sorry that I missed to answer the question.

I just mean, if I uploaded a e.g. 3000px wide portrait image, the image would take up a hugh amount of space on the page. Therefore, I probably would have added somethijng like `.name-of-img-class { width: 80%; height: auto; max-height: 80vh; object-fit: contain;}`
